### DISCUSSION TITLE
Using corrosion as a wrapper

### DISCUSSION THREAD

**Discussion by parthopdas** [corrosion-rs/corrosion#370] [General]
Is it possible to use corrosion as a simple wrapper for cargo executable packages? i.e. use corrosion just to binplace the rust exe built by cargo without linking it to C/C++ stuff.

I am trying to get cross platform Rust development working with Visual Studio 2022 e.g. write code on windows, build / debug on WSL2. VS2022 natively supports CMake.

So if I could use corrosion as a simple wrapper, then all the VS2022 infra works nearly for free.

**Comment by jschwe** [comment]
>  i.e. use corrosion just to binplace the rust exe built by cargo without linking it to C/C++ stuff.

Yes, that is possible (Assuming it is 100% rust code that doesn't _need_ to link with C/C++ code - in that case you would naturally get linker errors). Is this just a question or did you try and something didn't work?

> I am trying to get cross platform Rust development working with Visual Studio 2022 

Does Visual Studio support rust-analyzer? Personally I would find it quite cumbersome to write rust code without rust-analyzer or an equivalent language server.

**Comment by parthopdas** [comment]
yes VS supports rust-analyzer along with the other VS goodness ref. [rust-analyzer.vs](https://marketplace.visualstudio.com/items?itemName=kitamstudios.RustAnalyzer&ssr=false#overview).

i was thinking of packaging corrosion with that for the cross-plat experience. great artists steal and all that jazz ; )

> Yes, that is possible (Assuming it is 100% rust code that doesn't need to link with C/C++ code - in that case you would naturally get linker errors). Is this just a question or did you try and something didn't work?

i gave it a shot but could not figure out a way to not have c/c++. please can you give me a hint or two?

also you said both it is possible and that i would get linker errors - not sure if there was a typo.

**Reply by jschwe** [reply]
As long as the project is 100% Rust, i.e. you don't call any C/C++ code, so you don't need to link any C/C++ code, building should work just fine. For example a bin project that only contains

```rust
fn main() {
    println!("hello world"); // 100% Rust code
}
```

Should be able to be imported and built by corrosion just fine. 

If however, you do want to call some C++ code, like in the example below, then of course you would need to tell Rust which C/C++ library it should link to that contains `my_cpp_print_function`. When building Rust binaries, Rust will always invoke the linker.

```rust
extern "c" {
      fn my_cpp_print_function(name: *const c_char);
}

fn main() {
    unsafe { my_cpp_print_function("hello world\0") };
}
```

Now, if you want the linker to be invoked by CMake, then you probably don't want a Rust binary, you want a Rust `staticlib`, that you can then link into your C/C++ application with the regular mechanisms.

Does this answer your question?

**Comment by parthopdas** [comment]
it does, thank you, working on it. i'll get back.
